gax/src/main/java/com/google/api/gax/batching/v2/BatcherImpl.java

```diff
@@ -0,0 +1,216 @@
+/*
+ * Copyright 2019 Google LLC
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google LLC nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.batching.v2;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.util.concurrent.MoreExecutors.directExecutor;
+
+import com.google.api.core.ApiFuture;
+import com.google.api.core.ApiFutureCallback;
+import com.google.api.core.ApiFutures;
+import com.google.api.core.BetaApi;
+import com.google.api.core.InternalApi;
+import com.google.api.core.SettableApiFuture;
+import com.google.api.gax.rpc.UnaryCallable;
+import com.google.common.base.Preconditions;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Queues up the elements until {@link #flush()} is called; once batching is over, returned future
+ * resolves.
+ *
+ * <p>This class is not thread-safe, and expects to be used from a single thread.
```

Comment:
This class assigns callbacks with `directExecutor` (i.e. executor will be executed in same thread which resolved the future) and inside the callback we call this class's method `onBatchCompletion()` (in `finally` block of each callback). There is no guarantee that the callback thread and the thread instantiated this class will be same thread. It looks like this class has to be thread-safe.

Reply:
See comment above. Externally we consider this class not thread safe. But internally this class handles races between the single caller thread and multiple grpc threads for callbacks

Reply:
Maybe just make this class thread safe? People will be maintaining it, modifying etc. Even if everything works fine now, it will be easier to break things in the future. I do not insist, but it is my experience in general: classes are either not aware of threads whatsoever, or immutable or thread-safe. Not thread-safe (but uses threads stuff and is in general aware that it is used in concurrent), thread-compatible, thread-hostile - kind of time bombs, waiting until somebody uses them in a wrong way.

Reply:
Let's take it a step at a time. Thread safety will come up again when @rahulKQL gets to implementing timed flush. There are still at least 2 PRs coming up:
1a. autoflush based on element count
1b. autoflush based on request size
2. autoflush based on time

After we get an idea as to what is required for 2, we can have a much more informed discussion about external thread safety